File: collatz/collatz.c
#include <stdio.h>
long next(long x) { return x % 2 == 1 ? 3 * x + 1 : x / 2 ;}
long speed(long s) { return s == 1 ? 0 : speed(next(s)) + 1 ;}     
struct collatz {long n; long s;};
struct collatz max_speed(long n) {
    struct collatz c;
    c.s = 0;
    for (long i=1, s ; i <= n  ;  i++) {
        if ((s = speed(i)) > c.s)  { c.n = i; c.s = s; }
    }
    return c;
}


int main(void) {
    struct collatz c = max_speed(10000000);
    printf("%ld %ld\n",c.n, c.s);
}

File: collatz/collatz.rs
fn next(x:u64)->u64 { if x % 2 == 1 { 3 * x + 1 } else { x / 2 } }
fn speed(s:u64)->u64 { if s == 1 {0} else { speed(next(s)) + 1 } }     
fn max_speed(n:u64)->(u64,u64) {
    let (mut ms, mut mx)  = (0,0);
    for i in 1..=n {
        let s = speed(i);
        if s > ms  { 
            ms = s; 
            mx = i; 
        }
    }
    return (ms,mx);
}

fn main() {
    let (x,y) = max_speed(10000000);
    println!("{} {}",x, y);
}
